lightning/src/onion_message/messenger.rs

```diff
@@ -854,6 +854,27 @@ where
 		)
 	}
 
+	/// Forwards an [`OnionMessage`] to `peer_node_id`. Useful if we initialized
+	/// the [`OnionMessenger`] with [`Self::new_with_offline_peer_interception`]
+	/// and want to forward a previously intercepted onion message to a peer that
+	/// has just come online.
+	pub fn forward_onion_message(
+		&self, message: OnionMessage, peer_node_id: &PublicKey
+	) -> Result<(), SendError> {
+		let mut message_recipients = self.message_recipients.lock().unwrap();
+		if outbound_buffer_full(&peer_node_id, &message_recipients) {
+			return Err(SendError::BufferFull);
+		}
+
+		match message_recipients.entry(*peer_node_id) {
+			hash_map::Entry::Occupied(mut e) if e.get().is_connected() => {
```

Comment:
Would it be useful to hit the `ConnectionNeeded` path, too? Say a peer comes back online, the user gets the `OnionMessagePeerConnected` event, but the peer goes offline before `forward_onion_message` is called. When the user calls `forward_onion_message`, it errors and consumes the `OnionMessage`, which is now lost.

Reply:
Yeah, it should be rare that a peer is disconnected by the time we go to forward, but it is currently possible. As mentioned in the PR description, planning to address this in follow-up per the race condition TODO here: https://github.com/lightningdevkit/rust-lightning/issues/2950

Reply:
Not sure I understand your question, but it's up to the LDK user's event handling code to decide which peers it should store OM forwards for and which should be discarded. If the user decides an OM is worth storing and later calls `forward_onion_message` when the outbound peer comes back online, then we would go through the ping/pong flow to let them know when the OM is officially safe to delete.

Reply:
Oh, I think I misunderstood the proposed solution. I thought LDK would manage whether to forget the message (in `OnionMessenger`) based on whether the ping/pong completed. But, IIUC, the code in this PR would remain as is and LDK would instead surface some event when a peer comes online and has completed a ping/pong. Is that correct?

So, in the scenario I gave, the user would simply not store the message in the first place?

Reply:
> the code in this PR would remain as is and LDK would instead surface some event when a peer comes online and has completed a ping/pong. Is that correct?
> So, in the scenario I gave, the user would simply not store the message in the first place?

Yes to both!